### PR TITLE
sampler and span processor prototype implementations for consistent sampling

### PR DESCRIPTION
**Pull request by oertl**
**Description:**

This PR adds various Sampler and SpanProcessor prototype implementations for consistent sampling as defined by https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md and https://github.com/open-telemetry/opentelemetry-specification/pull/2047:
* **ConsistentSampler**: abstract base class of all consistent sampler implementations below
* **ConsistentAlwaysOffSampler**: see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md#always-off-sampler 
* **ConsistentAlwaysOnSampler**: see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md#always-on-consistent-probability-sampler
* **ConsistentComposedAndSampler**: allows combining two consistent samplers and samples when both samplers would sample
* **ConsistentComposedOrSampler**: allows combining two consistent sampler and samples when at least one of both samplers would sample, see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md#requirement-combine-multiple-consistent-probability-samplers-using-the-minimum-p-value
* **ConsistentParentBasedSampler**: see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md#parentconsistentprobabilitybased-sampler
* **ConsistentProbabilityBasedSampler**: see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-probability-sampling.md#consistentprobabilitybased-sampler
* **ConsistentRateLimitingSampler**: a rate limiting sampler based on exponential smoothing that dynamically adjusts the sampling probability based on the estimated rate of spans occurring to satisfy a given rate of sampled spans
* **ConsistentReservoirSamplingBatchSpanProcessor**: consistent reservoir sampling of spans, see documentation in the source code

EDIT: ConsistentReservoirSamplingBatchSpanProcessor was removed to make this PR smaller as requested by @anuraaga. It can now be found in https://github.com/dynatrace-oss-contrib/opentelemetry-java-contrib/tree/consistent-reservoir-sampling.

### CODE REVIEW

File: consistent-sampling/build.gradle.kts
Comment (anuraaga) [inline consistent-sampling/build.gradle.kts:9]:
```suggestion
  api("io.opentelemetry:opentelemetry-sdk-trace")
```

File: consistent-sampling/build.gradle.kts
Comment (anuraaga) [inline consistent-sampling/build.gradle.kts:10]:
```suggestion
  testImplementation("com.google.guava:guava")
```

Comment (anuraaga) [reply consistent-sampling/build.gradle.kts:10]: Though given it'd be very simple to replace the one usage with `HashMap` would recommend doing that instead

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/export/ConsistentReservoirSamplingBatchSpanProcessor.java
Comment (anuraaga) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/export/ConsistentReservoirSamplingBatchSpanProcessor.java:49]: Is this for tail sampling? It's not intuitive why we'd have a span processor doing sampling otherwise.

If possible I'd recommend holding off on this class until it's in the spec as documentation there would help a lot with understanding it, intuitively it seems unrelated to `ConsistentSampler`

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentAlwaysOffSampler.java
Comment (anuraaga) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentAlwaysOffSampler.java:12]: Our general pattern is to not make the implementations public and expose static factories on the `ConsistentSampler` itself

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentComposedAndSampler.java
Comment (anuraaga) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentComposedAndSampler.java:16]: Can you describe the difference in behavior vs composing the actual `SamplingResult` instead?

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentComposedOrSampler.java
Comment (anuraaga) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentComposedOrSampler.java:49]: Prefer to omit else statements when returning

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentParentBasedSampler.java
Comment (anuraaga) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentParentBasedSampler.java:15]:
```suggestion
 * A consistent sampler that makes the same sampling decision as the parent and optionally falls back
```

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentProbabilityBasedSampler.java
Comment (anuraaga) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentProbabilityBasedSampler.java:13]: Needs javadoc

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentSampler.java
Comment (anuraaga) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentSampler.java:26]: We want to reduce public API as much as possible in libraries. We can probably make it so this is the only public class, which has factories returning implementations, and nothing else would need to be public from what I can tell.

Comment (anuraaga) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentSampler.java:26]: Also favor composition over inheritance, so we should have a class that implements `Sampler`, containing a field that's a `ConsistentSampler` rather than talking through hierarchy

**Review by anuraaga — COMMENTED** [review]
Did a very quick skim - 3700 lines is too huge though, let's start with a PR that only adds `ConsistentProbabilityBasedSampler` please

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java
Comment (PeterF778) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java:96]: Any chances this algorithm can be adapted to support "worst-case" guarantee, such as a predefined confidence level that the specified rate will not be exceeded?

File: consistent-sampling/build.gradle.kts
Comment (oertl) [reply consistent-sampling/build.gradle.kts]: fixed in 6b4ba10eb16ba8392945f1e59b852e9b4f6be637

File: consistent-sampling/build.gradle.kts
Comment (oertl) [reply consistent-sampling/build.gradle.kts]: fixed in 6b4ba10eb16ba8392945f1e59b852e9b4f6be637

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java:96]: To calculate such a "confidence level", a model is needed that describes when the spans occur. The simplest model is probably the Poisson process, which assumes that the rate is constant and that all spans are independent. However, both assumptions are rarely true in practice. The rate is usually not constant, and the occurrence of spans is also not independent (consider, for example, a trace that always calls a service twice, so spans are always generated in pairs). 

We also need to specify how the rate limit is defined. Usually this is done in terms of a time interval. E.g., a maximum of 200 spans in each 1-minute interval.  

When everything is well defined, the model, the rate limit definition, and the adaptive sampling algorithm, the expected frequency of rate limit violations can be calculated or simulated. I assume this is what you are ultimately interested in?

Comment (PeterF778) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java:96]: Yes, I believe an algorithm that gives some kind of confidence level is a good compromise between a purely probabilistic approach and any algorithm based on "leaky bucket" approach. I believe it can be made almost-unbiased and almost-rate-limiting.
BTW, your algorithm, when faced with spans coming in pairs (at almost the same point of time) can reduce the sampling probability for the second span significantly.

Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java:96]: > BTW, your algorithm, when faced with spans coming in pairs (at almost the same point of time) can reduce the sampling probability for the second span significantly.

Yes, this is true if the parameter for the adaptation time parameter is chosen small (in the order of magnitude of the time span between subsequent pairs or less).

This algorithm was motivated by a scenario where the span rate is suddenly increased. In this case, the algorithm immediately starts to decrease the sampling probability. In contrast, an adaptive sampler that adjusts the sampling rate periodically (e.g., based on the estimated rate of the previous period) would take much longer and the desired sampling rate would be exceeded much further.

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentComposedOrSampler.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentComposedOrSampler.java]: addressed in 000ffea29389658bab35f838b7cdd185e74bc97d

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentProbabilityBasedSampler.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentProbabilityBasedSampler.java:14]: addressed in 9258ef4b029ba527179a7c1e6f991873849db5b5

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentAlwaysOffSampler.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentAlwaysOffSampler.java]: addressed in f7911c35c4c502b205493b02e10e5e441d47a00b

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentComposedAndSampler.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentComposedAndSampler.java:16]: please see added javadoc in b3459f9e79d1a69c17e4cdf7aeb811801b692b6b

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentSampler.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentSampler.java]: > We want to reduce public API as much as possible in libraries. We can probably make it so this is the only public class, which has factories returning implementations, and nothing else would need to be public from what I can tell.

addressed in f7911c35c4c502b205493b02e10e5e441d47a00b

Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentSampler.java]: > Also favor composition over inheritance, so we should have a class that implements `Sampler`, containing a field that's a `ConsistentSampler` rather than talking through hierarchy

I think inheritance is appropriate in this case. First, `ConsistentSampler` is a `Sampler`. And secondly, the construction of a composed consistent sampler that is the AND/OR combination of 2 consistent samplers would be more difficult otherwise.

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/export/ConsistentReservoirSamplingBatchSpanProcessor.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/export/ConsistentReservoirSamplingBatchSpanProcessor.java:49]: > Is this for tail sampling? It's not intuitive why we'd have a span processor doing sampling otherwise.

No, this is different from tail sampling, where a sampling decision is made for the entire trace after all spans of a trace have been sampled. The ConsistentReservoirSamplingBatchSpanProcessor can be used to implement rate-limited sampling of spans with a hard limit. Since the sampler interface requires an immediate sampling decision, it is not straightforward to ensure that the rate limit is never exceeded, if the number of spans suddenly increases. For example, the ConsistentRateLimitingSampler may exceed the desired span rate in such cases. By delaying the sampling decision, as the ConsistentReservoirSamplingBatchSpanProcessor does, a hard limit can be achieved while giving all spans a fair chance to be sampled.

Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/export/ConsistentReservoirSamplingBatchSpanProcessor.java]: > If possible I'd recommend holding off on this class until it's in the spec as documentation there would help a lot with understanding it, intuitively it seems unrelated to ConsistentSampler

I have removed the ConsistentReservoirSamplingBatchSpanProcessor from this PR (see dd3df4ed697e968ee822228a92ad5aba58c32a6a)

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java
Comment (spencerwilson) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentRateLimitingSampler.java:23]: Looking at Wright's method (equations 2.5, 2.7 [here](http://quantitative.cz/wp-content/uploads/2018/09/improved-holt-method-for-irregular-time-series-tomas-hanzak-2008.pdf)), there seem to be some differences from what you've implemented. E.g., there are no `1 - decayFactor` terms in this implementation, whereas 1 - alpha appears in Wright. Is this an intentional deviation?

Maybe because the setting in OTel sampling is just different than that of forecasting? My understanding of this implementation is: we're trying to estimate a rate, so you need to maintain both an exponentially smoothed span count and an exponentially smoothed window duration in order for their ratio to be a meaningful: an exponential moving _rate_ (in analogy to exponential moving average). Remarkably, I can't find any literature on "exponential moving rate".

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java
Comment (jmacd) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java:187]:
```suggestion
```

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java
Comment (jmacd) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java:194]:
```suggestion
```

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java
Comment (jmacd) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java:210]:
```suggestion
```

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java
Comment (jmacd) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java:234]:
```suggestion
```

**Review by jmacd — APPROVED** [review]

**Comment by trask** [comment]
Hi @oertl! As a spec'd component (even if still experimental), I think a better place for this would be in the core java repo, under [sdk-extensions](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions) (e.g. alongside the jaeger remote sampler).

We can definitely merge it here first, but if we are planning to move it to the core repo it may be worth moving it sooner than later to avoid user confusion when the groupId changes in the move (from `io.opentelemetry.contrib` to `io.opentelemetry`).

@open-telemetry/java-maintainers what is your preference on how to proceed? I'm happy to support any path forward. It's not a big deal to move it later if you prefer for it to incubate here at least initially. We have renamed many many artifacts already 😭.

**Comment by oertl** [comment]
Hi @trask! I would prefer to merge it here first, since some sampler implementations are not yet described in the spec.

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java
Comment (trask) [inline consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java:27]: can this use empty list instead of null?

**Review by trask — COMMENTED** [review]
thx @oertl!

I'm hoping you are able to be a component owner for this? if so, can you update https://github.com/open-telemetry/opentelemetry-java-contrib/blob/main/.github/component_owners.yml?

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/state/OtelTraceState.java:203]: The proposed condition is not equivalent. However, I noticed that this condition also passes the unit tests. Therefore, I will add some more test cases, to make clear that the proposed condition is not an option.

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/util/RandomUtil.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/util/RandomUtil.java:50]: added some lines to the javadoc to explain this algorithm

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/util/RandomGenerator.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/util/RandomGenerator.java]: addressed in edac8a9a517ed79d772d80b758bc56a6e5d7397a

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/util/RandomUtil.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/util/RandomUtil.java]: addressed in edac8a9a517ed79d772d80b758bc56a6e5d7397a

File: consistent-sampling/src/main/java/io/opentelemetry/contrib/util/RandomGenerator.java
Comment (oertl) [reply consistent-sampling/src/main/java/io/opentelemetry/contrib/util/RandomGenerator.java]: addressed in edac8a9a517ed79d772d80b758bc56a6e5d7397a

**Comment by oertl** [comment]
@trask, thank you very much for your valuable suggestions! I hope I have not overlooked anything and have addressed all your points.

**Review by trask — APPROVED** [review]
@open-telemetry/java-contrib-approvers let me know if anyone is planning to review this further, otherwise I will merge before making the upcoming release, thx!

**Review by anuraaga — COMMENTED** [review]
It looks like `RandomGenerator` can be moved into the `samplers` package and made package private? We can have package-private factories accepting it for use in unit tests only in this repo, but let's keep the public API using just a default threadlocalrandom based generator without user configuration.

For `OtelTraceState` it also looks like it can be made package private. If you expect other packages needing the utility methods, then it's also ok to move into `io.opentelemetry.contrib.samplers.internal` but suspect we don't need to do more than package private at least at this time.

Note that we cannot have packages present in multiple artifacts in our codebase. So currently the generic sounding packages `io.opentelemetry.contrib.state` and `io.opentelemetry.contrib.util` could cause problems in the future. The above internalizing suggestions will hopefully fix that naturally.

**Review by anuraaga — APPROVED** [review]
Thanks!

**Comment by oertl** [comment]
@anuraaga, thanks for your suggestions. I addressed them in da4201893e6ea2674fb4b4f5f7a5337de8d1d23f

**Comment by trask** [comment]
thanks @oertl 🎉

**Pull request merged**
